### DISCUSSION TITLE
MIDI keyboard connection.

### DISCUSSION THREAD

**Discussion by aastam** [GrandOrgue/grandorgue#1186] [Help]
I have a yamaha PSR 220 keyboard and I have bought a midi-usb cable to connect it to my laptop. I have downloaded grand orgue and currently I am with the demo organ. I don't understand how to connect my keyboard to GrandOrgue in order to play. I click listen for event but nothing happens. Your help will be very much appreciated.

**Comment by kerkovits** [comment]
For me, the usual reason for it is that the keyboard must be powered on **a few seconds before** launching GrandOrgue. It would be nice if GO could connect to newly connected MIDI devices on-the-fly, but as far as I know, it is not possible due to performance reasons.

**Reply by FelipeFTN** [reply]
Hello, @oleg68!
Is that information true?

I'm having a bit of a headache with my MIDI USB. Sometimes my keyboard turns off automatically and GrandOrgue lost connection with the MIDI Keyboard. Is there some way to get that connection back, without restarting the entire GrandOrgue application?

I found that there is a button "Reload" under File->Reload tab. But by clicking it, it just reloads the organ window, not the GO app; and the keyboard doesn't recognize back unless I close the GO and open it again, which can take some extra time.

Is there some workaround? Perhaps the "Reload" button could do this with some new feature or something? If it's possible, I'm able to open a PR as a new feature!

Thanks a lot. 

PS: Seems like the 'panic' button, under 'Audio/MIDI' does exactly that.

**Comment by aastam** [comment]
Ok thanks.
Could you please describe for me the process to connect it properly?

Στις Τρί, 30 Αυγ 2022, 09:08 ο χρήστης kerkovits ***@***.***>
έγραψε:

**Comment by rousseldenis** [comment]
> Ok thanks. Could you please describe for me the process to connect it properly? Στις Τρί, 30 Αυγ 2022, 09:08 ο χρήστης kerkovits ***@***.***> έγραψε:
> […](#)
> For me, the usual reason for it is that the keyboard must be powered on *a few seconds before* launching GrandOrgue. It would be nice if GO could connect to newly connected MIDI devices on-the-fly, but as far as I know, it is not possible due to performance reasons. — Reply to this email directly, view it on GitHub <[#1168 (comment)](https://github.com/GrandOrgue/grandorgue/issues/1168#issuecomment-1231187204)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/A2ZTRR4QACYMHH74WDPAWTTV3WQNFANCNFSM576CZHMQ> . You are receiving this because you authored the thread.Message ID: ***@***.***>

Could you give:

- Your operating system
- The USB cable model
- A screenshot of File > Settings > MIDI tab ?

Have you looked there ? https://github.com/GrandOrgue/grandorgue/wiki/Broken-MIDI-Adapters

**Comment by jason-barba** [comment]
Are you sure that your MIDI-to-USB converter is working properly. My first one was faulty wired in its connectors.

**Comment by aastam** [comment]
Hello,
My operating system is laptop Dell Inspiron 3511, running windows 11 home
and it is 64 bit. My keyboard is a Yamaha PSR-220. The midi usb cable is
probably this one:
https://www.electronicaeshop.eu/en-gb/usb-to-midi-keyboard-interface-cable-cord-adapter-converter-oem-en-gb.html
.
I don't know how to find the midi tab in order to send you the screenshot
of the file.
Although I am sending you a photo of the indicating lights on my cable.
Only the light "OUT" is indicating.
Your help would be very much appreciated.
[image: 302443543_5329861300446345_3401327251491628576_n.jpg]

Στις Τρί 30 Αυγ 2022 στις 11:40 π.μ., ο/η Denis Roussel (ACSONE) <
***@***.***> έγραψε:

**Comment by oleg68** [comment]
> I don't know how to find the midi tab in order to send you the screenshot of the file.

![midi-settings](https://user-images.githubusercontent.com/11586506/188314015-e1bc064f-7bd8-4d0e-a9da-cebc5dae8807.png)

**Comment by oleg68** [comment]
@classicpianist78 We cann't help you without the screenshot of File->Settings->MIDI Devices

**Comment by madbrain76** [comment]
I'm experiencing the same issue on Linux, with a Roland UM-880, and Roland VK-88 organ connected to MIDI 3 IN.

If I unplug the UM-880, and plug it back in, the Alsa MIDI connections disappear temporarily, then reappear. The problem is that Grandorgue does not notice the disconnect, and thus does not try to reconnect. I used codex to come up with a workaround that uses the aconnect cli tool to manually force the connection between Alsa MIDI and Grandorgue to be re-established. This workaround needed to have knowledge of the exact MIDI configuration I used as arguments on the CLI, and thus is very brittle. If I ever wanted to change the device or ports, the workaround would no longer work. For now, I have this running as a systemd user service in my lxc container under Proxmox, with GO running under xfce inside a VNC virtual display.

Codex says a proper fix is for GO to subscribe to the Alsa connect/disconnect events, and notice when the configured MIDI ports disappear/reappear, and automatically reconnect the ports. I'm having Codex pull the GO source code and build it in that container. I will let it take a stab at implementing the fix, and test it myself. However, I don't know anything about Alsa or the Grandorgue source code myself, as I haven't worked personally with these projects before. It may be best for one of the GO maintainers to take a look. I'm happy to submit a PR if the fix works for me, but I don't know the process for getting it accepted upstream.

**Comment by madbrain76** [comment]
FYI, this is the workaround Codex came up with, which works when I hotplug my UM-880. Not sure if it has any applicability to the OP who is running Windows.

```
madbrain@grandorgue:~/.config/systemd/user$ cat grandorgue-midi-hotplug.service 
[Unit]
Description=Reconnect GrandOrgue to the UM-880 after MIDI hotplug
After=vncserver.service

[Service]
Type=simple
ExecStart=/home/madbrain/.local/bin/grandorgue-midi-hotplug
Restart=always
RestartSec=2

[Install]
WantedBy=default.target
madbrain@grandorgue:~/.config/systemd/user$ cat /home/madbrain/.local/bin/grandorgue-midi-hotplug
#!/bin/sh

# Keep GrandOrgue's configured UM-880 inputs connected after USB hotplug.
# GrandOrgue creates one ALSA client per configured input: the first is the
# MIDI Through port, followed by UM-880 MIDI 1-8 and the control port.

lock_file="/run/user/$(id -u)/grandorgue-midi-hotplug.lock"
exec 9>"$lock_file"
/usr/bin/flock -n 9 || exit 0

while :; do
    midi_graph=$(/usr/bin/aconnect -l 2>/dev/null || true)
    um_client=$(printf '%s\n' "$midi_graph" | /usr/bin/awk -F '[: ]+' "/^client [0-9]+: 'UM-880'/ { print \$2; exit }")
    go_clients=$(printf '%s\n' "$midi_graph" | /usr/bin/awk -F '[: ]+' "/^client [0-9]+: 'GrandOrgue'/ { print \$2 }")

    if [ -n "$um_client" ]; then
        set -- $go_clients
        if [ "$#" -ge 10 ]; then
            # Skip the first GrandOrgue client, which belongs to MIDI Through.
            shift
            um_port=0
            for go_client in "$@"; do
                [ "$um_port" -ge 9 ] && break
                case "$midi_graph" in
                    *"Connecting To: $go_client:0"*)
                        ;;
                    *)
                        /usr/bin/aconnect "$um_client:$um_port" "$go_client:0" >/dev/null 2>&1 || true
                        ;;
                esac
                um_port=$((um_port + 1))
            done
        fi
    fi

    sleep 2
done

```

**Comment by madbrain76** [comment]
Seems like the proper fix for this would be in the Rtmidi Alsa backend. Somebody already reported the issue 7 years ago, but it wasn't fixed. It looks like the problem was that disconnect/reconnect is not equally available across all operating systems and MIDI APIs. I will continue my investigation in that project.

https://github.com/thestk/rtmidi/issues/191

**Reply by Psychlist1972** [reply]
It can be done in Windows; it's just not implemented in rtMIDI

And it is a bit hairy when devices don't have serials, or they get disconnected from one USB port, but plugged into another, which is actually behind a hub, or something else with its own identity, or the customer plugs in another device of the same make/model, but completely different internal configuration and so not actually the same device.

Pete
Microsoft

**Reply by madbrain76** [reply]
Thanks. My UM-880 does report a serial number. The patch I submitted for the rtmidi ALSA backend takes that into account. Identical serial is a definitive signal that it is the same device being reconnected. I can see how devices without serials and re-appearing on different buses might be ambiguous. I think for those cases, not attempting an automatic reconnection would be fine.  However, if a serial-less device reappears on the same bus, same hub, and same port, IMO, it should be assumed to be the same device, and automatic reconnection should work.

**Comment by midimusic** [comment]
Surely this is trying to solve a problem that does not really exist. If your keyboard is connected to GO with a midi to USB cable the port is created when the computer is powerd on - it makes no difference whether the keyboard itself is powered on or not. Thus the port is always availabe to GO when it is loaded and run.

Even if the keyboard is turned off or disconnected the port will not be lost. It will only be lost if you unplug the USB connection from the computer and this is not a very sensible thing to do.

csw900

**Reply by madbrain76** [reply]
> Surely this is trying to solve a problem that does not really exist. If your keyboard is connected to GO with a midi to USB cable the port is created when the computer is powerd on - it makes no difference whether the keyboard itself is powered on or not. Thus the port is always availabe to GO when it is loaded and run.
> 
> Even if the keyboard is turned off or disconnected the port will not be lost. It will only be lost if you unplug the USB connection from the computer and this is not a very sensible thing to do.

You are assuming that the MIDI to USB cable is bus-powered. That is not the case for my UM-880. It is rack mounted, and externally powered. It features a physical power button on the front. I do not want to keep it powered perman tly. No physical cable disconnection is required for the port to disappear from the host computer.

In the situation I ran into, the UM-880 was powered off when my host Proxmox host rebooted after I made some BIOS changes. The lxc started Grandorgue. Hours later, I powered up the UM-880 and the VK-88. I hit keys on the keyboard, and did not hear any sound. That is a problem that actually exists for me, and I would not be trying to fix it if it did not.

The host PC itself will stay on permanently, because it runs other services such as my pfSense router VM, Home assistant, and various other services. The MIDI interface will not as it's not needed permanently.

Obviously, workarounds exist - restarting Grandorgue, or the script I posted that does the reconnection. I just don't think they should be needed. When I want to play the organ, I just want to hit the keys. I don't want to even think about the computer. I don't want to be forced to turn on my desktop and login to proxmox to reboot the Grandorgue container, or use VNC viewer, login to it, and restart the Grandorgue GUI. My use case is that Grandorgue runs in the background at all times, and I use it as an always-on sound module. I only want to interact with the software if I'm changing software settings, such as loading another organ, similar to the way I interact with my Fantom-XR MIDI module - if I'm not changing the sounds on it, I just press the power button and never touch any of the buttons on the front.

**Comment by madbrain76** [comment]
Additionally, the MIDI interface not being connected at power-up time is not the only disconnect situation. I power up the MIDI interface whenever I play, and power it off whenever I stop playing, with its power button. Without this fix, that means having to login to the computer to restart Grandorgue every time I want to play. That is a very annoying thing to be required to do every time one plays. A non-digital organ does not require that. Neither do my acoustic pianos or harpsichords. This is about making my organ a seamless, easily playable instrument.

**Comment by midimusic** [comment]
@madbrain76 
You obviously now understand the situation with GO ports, and bearing in mind the way you use your system it is likely that using  the Roland UM880 is not the best hardware to connect keyboards to your computer. You need something that is either powered from the USB connection or something you are happy to leave on all the time.

You need to do a bit of research to find the best solution - I do not think that it is a good idea to try and modify GO so it automatically connects to ports which have been closed. There could be nasty side effects such as connections to the wrong ports.

csw900

**Reply by madbrain76** [reply]
Actually, the UM-880 is one of the best MIDI interfaces ever made - an 8x8 matrix with a lot of flexibility. I don't believe a bus-powered equivalent exists. And from a power consumption standpoint, it is still better to be able to turn it off.

I don't think Grandorgue itself needs any modification to handle auto-reconnection. The rtmidi library that it statically links with does, though. I have created a patch for it, and sent a PR.

https://github.com/thestk/rtmidi/pull/374

**Comment by willeke1234** [comment]
Some GO users do not have a dedicated GO computer and use the equipment they already have.

I have to connect my headphones and Bluetooth MIDI devices before I start GO, but I usually forget something. So I start GO and start loading a sample set. In the meanwhile I connect the Bluetooth MIDI devices. Once everything is connected to my computer, I press the panic button to have GO rescan the devices, which may have side effects. I would prefer automatic connection, but RtMidi does not support this. In GO, it wouldn't be that difficult; only a trigger is needed and a fresh scan of the MIDI devices, just like the Panic button and the Settings window do.

Perhaps a workaround is a MIDI routing app that stays open, automatically connects, and forwards to permanent virtual MIDI ports.

**Reply by madbrain76** [reply]
Yes, I understand most GO users won't have a dedicated GO computer, and even those do may not leave it on 24/7. For me, this is about set and forget - I press the power button on the organ and just get to play it.

I created a patch for rtmidi to handle automatic reconnection with the Alsa backend. See https://github.com/thestk/rtmidi/pull/374 .

It is extensible, and transparent to Grandorgue, without the need to open the GUI. Not one line of Grandorgue source code is changed. I didn't implement reconnection in other backends, as I don't have the software & hardware combinations to test and fix them all. Other backends continue to work the same way they do today, ie. without automatic reconnection. If my patch is accepted upstream, I could test Windows and contribute the code for it, also.

**Reply by willeke1234** [reply]
> Other backends continue to work the same way they do today, ie. without automatic reconnection.

macOS/Core MIDI already has automatic reconnection.

**Reply by madbrain76** [reply]
> > Other backends continue to work the same way they do today, ie. without automatic reconnection.
> 
> macOS/Core MIDI already has automatic reconnection.

Thanks ! I did not know that. After researching it, I found that the automatic reconnection logic is not in the RtMidi backend, but in MacOS itself. Thus, the RtMidi backend for macOS/ CoreMIDI does not need any additional code change.

**Comment by oleg68** [comment]
@madbrain76 Have you tried to use jack instead of alsa? Does this promlem exist there?

**Reply by madbrain76** [reply]
> @madbrain76 Have you tried to use jack instead of alsa? Does this promlem exist there?

As I understand, Jack is a layer that sits between Alsa and Grandorgue. Thus, it would suffer from the suffer the same problem as using Alsa directly, and would not resolve the automatic disconnection/reconnection issue.

**Reply by oleg68** [reply]
Have you tried it?

**Reply by madbrain76** [reply]
Not, yet, because it seems quite involved, and the research I have done says it doesn't claim to solve hotplug issues. Do you have any reason to believe that's wrong ?

Here is what I found :

ChatGPT :

_**JACK itself does not automatically solve USB MIDI hotplug/reconnection.

If the UM-880 is exposed to JACK through an ALSA→JACK MIDI bridge such as a2jmidid, then:

UM-880 disappears → its ALSA sequencer ports disappear.
The corresponding JACK MIDI ports disappear too.
UM-880 comes back → a2jmidid creates JACK ports for it again.
Existing JACK connections to the old ports are not inherently restored. They're new JACK ports.

So plain JACK + a2jmidid does not give GrandOrgue a permanently stable endpoint across a UM-880 power cycle. You still need something that reconnects the ports, or a persistent proxy/patchbay layer.

If your goal is specifically:

GrandOrgue opens MIDI once, UM-880 can be switched off and back on, and GrandOrgue receives MIDI again without doing anything.

then switching GrandOrgue from ALSA MIDI to JACK MIDI alone does not guarantee that behavior.

A persistent ALSA proxy or fixing RtMidi/GrandOrgue's reconnect behavior addresses the actual problem.**_

Gemini :

**_The explanation you shared is accurate.

ALSA Sequencer vs. JACK MIDI Dynamics

    Port Lifecycle: When a USB MIDI device like the Roland UM-880 power cycles, its underlying system device file and ALSA sequencer client ID disappear. When a2jmidid acts as the bridge, it dynamically destroys the dead JACK MIDI ports.

    Re-enumeration: When the device powers back on, ALSA assigns it a new client ID. a2jmidid detects the new device and exposes brand-new JACK ports.

    Connection Loss: Standard JACK graph connections do not persist across destroyed ports. Unless an external patchbay manager (like aj-snapshot, PipeWire graph managers, or custom udev/scripting) dynamically re-links the newly created ports, the audio application remains disconnected.

GrandOrgue's Architecture
GrandOrgue relies on backend MIDI libraries (like RtMidi) to listen to specific port endpoints. If a device resets:

    With direct ALSA MIDI: GrandOrgue loses its file descriptor handle to the sequencer port. Unless GrandOrgue or RtMidi explicitly implements auto-reconnect polling, it will not receive events when the device reappears under a new client ID.

    With standard JACK MIDI: GrandOrgue’s JACK input port stays open, but the upstream connections from a2jmidid break and are not auto-restored.

Effective Solutions

    Persistent ALSA Proxy (e.g., midisnoop or Virtual ALSA RawMIDI/Sequencer Ports): Connect GrandOrgue to a static virtual MIDI port that never dies. Use an auto-connecting background daemon to route the UM-880 into that static virtual port whenever the UM-880 becomes available.

    PipeWire / WirePlumber: If running PipeWire instead of traditional JACK, WirePlumber rules can automatically re-establish saved connections when known MIDI hardware re-appears.

    Connection Managers: Running a daemon like aj-snapshot -a alongside JACK continuously monitors the graph and restores lost connections as soon as new hardware ports matching past names appear.

The statement correctly highlights that switching GrandOrgue's audio/MIDI driver from ALSA to plain JACK + a2jmidid without a persistence layer does not solve the USB hotplug problem._**

Perplexity - could not copy/paste to work, so here is a link : https://www.perplexity.ai/search/f50d2e1a-078f-4d39-bf39-6da88a7f613c .

Claude :

**_This checks out — no corrections needed on the technical substance.

A couple of things worth flagging, not because they're wrong but because they're worth being precise about:

- **a2jmidid port renaming behavior**: worth confirming which a2jmidid version/mode you're using. Older versions create new port names on reconnect if the ALSA client ID changes (which it usually does on USB replug — new client number assigned). Some patched/forked versions try to preserve names, but the underlying JACK port identity still changes even if the *name* is the same — JACK connections are keyed by port, and a new port object breaks existing connections regardless of whether it happens to get the same name back. So your point 4 is right either way, but it's worth being clear that even a "same name" reconnect doesn't save you — you'd need something actively watching and reconnecting (like `a2j_control` scripting, or a patchbay like `jack_matchmaker`, QjackCtl's patchbay-with-autoconnect, or Carla's patchbay persistence).

- **RtMidi/GrandOrgue ALSA reconnect**: this is real and worth noting explicitly — GrandOrgue's ALSA MIDI backend (via RtMidi) generally does not watch for ALSA sequencer client hotplug events after initial device enumeration. So even without JACK in the picture, unplugging/replugging the UM-880 while GrandOrgue is running against ALSA MIDI directly typically also does *not* self-heal without a restart of the MIDI subsystem in GrandOrgue (or the whole app). That's consistent with your framing — it's not a JACK-specific gap, it's that neither layer (ALSA-direct or JACK) has hotplug-aware reconnect logic baked in by default.

So the overall conclusion — that you need either a persistent reconnecting proxy/patchbay, or an actual fix/config in GrandOrgue's MIDI handling — is correct. Nothing here is misstated._**

**Comment by davidgritter** [comment]
for linux users it is possible to use the qjackctl (x11) or qpwgraph (pipewire)  patchbays to re-make these connections if the midi or audio device is temporarily lost.  However, at least for midi devices they need to be active when grandorgue is started or when you hit panic so that grandorgue will create  the necessary midi input and output ports

**Reply by madbrain76** [reply]
> for linux users it is possible to use the qjackctl (x11) or qpwgraph (pipewire) patchbays to re-make these connections if the midi or audio device is temporarily lost. However, at least for midi devices they need to be active when grandorgue is started or when you hit panic so that grandorgue will create the necessary midi input and output ports

Thanks. A panic is not exactly user-friendly behavior :-) And I do require things to work even if Grandorgue is started with my UM-880 powered off. Also, the UM-880 itself is a hardware MIDI patch bay, and adding another patch bay seems like the wrong approach.

**Reply by oleg68** [reply]
> for linux users it is possible to use the qjackctl (x11) or qpwgraph (pipewire) patchbays to re-make these connections if the midi or audio device is temporarily lost.

It would require GrandOrgue to expose a pure jack in-out connection not related to a certain MIDI device, like GrandOrgue Native Jack audio port.

**Comment by midimusic** [comment]
@madbrain76
I have now had a chance to read the User Handbook for your Roland UM880 and now think that you are not making the best use of it in relation to how you want to use GO.

It has 8 midi inputs and a MERGE function which allows up to 8 inputs to be merged into one of the 8 midi outputs. If you now connect this midi output via a midi to USB cable it will provide the input to GO using a midi port which will NOT BE LOST when you turn off the UM880.

Of course the midi inputs will need to each be on a different midi channel so GO can separate them. But this must also be the case with the built in USB output.

Thus your problem of losing the midi ports when the UM880 is turned off will be immediately solved and we can all forget about questionable changes to GO to accommodate a problem which does not exist.

csw900

**Reply by madbrain76** [reply]
I wouldn't be discussing a problem if it did not exist. If there was no problem, you wouldn't suggest I buy additional hardware, which is not something I'm interested in doing. I would like to improve the software, which is why I'm contributing my time and code.

**Comment by midimusic** [comment]
Here in simple terms is why Automatic Reconnection of midi ports cannot be reliably achieved.

Assume you have a system with two midi ports both with the same name (not at all uncommon) and both are open and active. Assume now that both are disconnected. This will close both of them. Assume that ONE of them is now reconnected - there is no way with the current API's, to determine which one of the two has been reconnected. Thus it is not possible to reliably re-open it.

According to Psyclist1972 above Microsoft has now solved this problem but it is not in the API's used by GO.

csw900

**Reply by madbrain76** [reply]
Each USB device is connected through a specific USB bus/controller, hub, and port, and has USB hardware properties such as vendor and product IDs and, where provided by the device, manufacturer, product, and serial-number information. Windows, Linux, and macOS expose this information, as well as USB connection/topology information, to applications, with various system APIs. This information can be utilized to make automatic disconnection and reconnection work reliably. It sounds like Apple has made this work transparently in its system APIs already, so clearly it is possible. For Windows, Psyclist1972's message implies that it is possible too. It is possible on Linux too, with the patch I created for rtmidi.

**Reply by Psychlist1972** [reply]
Not entirely, though. No solution here is 100%.

I was doing VID/PID/Serial matching at first and then had to pull it because the "I have more than one of the same device" is more common than one might think.

1. Most MIDI devices don't have a serial number. That's why I wrote up this article and got at least some manufacturers to agree to prioritize this for MIDI 2.0 devices going forward. https://devblogs.microsoft.com/windows-music-dev/the-importance-of-including-a-unique-iserialnumber-in-your-usb-midi-devices/ . Many MIDI 1.0 devices with inexpensive controllers just don't have the ability to write even a random number to nvram so they can't create a serial, and no manufacturer wants to add a serial number burn step to their production lines.

2. There is a surprisingly large number of devices out there with a hard-coded non-unique serial number. That's even worse than not having any serial at all because you don't know it's non-unique until you get the second one connected.

3. Not really a concern here, but for underlying libraries like rtMIDI, it is. Some USB devices, when they reboot into bootloader mode as part of a sysex-driven firmware update (a very common use for rtMIDI), actually change their PID. I've run into this already. All the auto-reconnect logic goes out the window, and apps start matching on names again, which is messy.

4. Also important for libraries: USB is not the only type of device connection. It's the one with the most problems like this, but Windows, for example, supports three types of virtual / app-to-app connections, Network MIDI 2.0, and Bluetooth MIDI 1.0, all of which have their own complexities. Like if you want to use an iPhone as a BLE-connected controller, and you don't pair it, it generates a random address each time.

Minus the bootloader thing, Windows does its best to keep the device ids static, as does macOS. (I assume ASLA does as well.) But in those cases, it's like an 80/20 solution, not a 100% solution. When you have more than one of the same device, which is super common for DJs/producers and pad controllers, it throws it all out the window when the devices don't have serial numbers.

The layer best equipped to deal with this is the operating system. Anything a library throws on top of that just constrains the OS from moving forward in the future when better solutions become possible.

That said, I'm not taking a side in the discussion. Just pointing out the complexity.

Pete
Microsoft

**Reply by madbrain76** [reply]
Thank you for the very helpful comment and article. That's certainly food for thought.

1. I wasn't aware of MIDI 2.0 . My USB MIDI devices, a UM-880, and Fantom XR (currently on loan), are ancient and definitely MIDI 1.0 .
2. The possibility of a device reporting a non-unique serial number is not something I anticipated. I am not sure how one could deal with that. I only have one UM-880 to check. Maybe the per-device behavior could be crowd-sourced, to see how serial actually works across USB MIDI devices ? And of course it would be interesting to compare the reported serial between different operating systems. I may spin a Windows VM on my Proxmox server to see what that looks like vs the host.
3. That's pathological ! But in that case, couldn't the auto-reconnect logic just punt on it ?
4. didn't know about BLE MIDI, now I have to look into this more just to play with it. I use Android though, not iOS. I don't personally have a vested interest in those other devices, except MUNT, the virtual MT-32 emulator. I haven't set it up in my Linux lxc yet. I probably wouldn't be hitting the keys to play its sounds, though.

Unfortunately, your assumption about ALSA trying its best in this situation is incorrect. I don't want to paste more LLM content here, so I'm sharing the link. https://chatgpt.com/s/t_6aa359c9a53c8191bb370abad71a7786

I'm not sure if trying to fix this at the Alsa level would be better than at the MIDI level. Probably a lot harder to get accepted upstream, though.

**Reply by Psychlist1972** [reply]
Thanks. I generally agree with what ChatGPT had to say. 

In Windows, we have another id layer in between what Linux/ALSA seems to have.

**Port Number:** Pretty much never the same twice. This sounds like the ALSA sequencer id. In the Windows case, is a number starting from 0 and increasing, with no gaps allowed, so a device removed in the middle recalculates the ids of all the ones that were after it. It's the WinMM approach that hasn't changed since 1993, and is useless as any sort of identification between connections. 

**The Port's device interface id.** This is a longer string like `\\?\swd#mmdevapi#midiu_ksa_6154580289096425992_1_12#{6dc23320-ab33-4ce4-80d4-bbb3ebbf2814}`  or `\\?\swd#mmdevapi#midiu_blemidi_partykeys_d0cf1351575d_1_0#{6dc23320-ab33-4ce4-80d4-bbb3ebbf2814}` . This is the primary thing the MIDI stack creates and operates on. It's our primary identifier. We have three different styles of those: MIDI 1.0 Source / Input, MIDI 1.0 Destination / Output, and UMP Endpoint (Bidirectional)

Then you have the **parent device**, and that is what has the VID/PID/Serial associated with it in the case of USB. It also has some additional information based on where it is plugged in if it doesn't have a serial, much like the examples in your research. 

The response also articulated part of why this is an 80/20 rule:
- If the device is plugged in to a different port or hub, the OS has to decide if it thinks it's the same device, or a new one
- If you have two USB-identical devices (doesn't matter if they have different patches or setup) and you swap which USB port they are on, the OS has no idea that they were swapped. This is really only safe to do when the controllers or synths are identical in setup.
- iSerialNumber, as mentioned. In Windows, the second time we see one with the same serial, we have to fall back to non-serial-number approach to identifying the device.

FWIW, `iSerialNumber` is part of the device's USB descriptors. There's not going to be any difference there between Windows/Linux/macOS/Android/etc.. You can check it on other operating systems, but it's unlikely to provide any clarity.

Pete
Microsoft

**Comment by willeke1234** [comment]
>And I do require things to work even if Grandorgue is started with my UM-880 powered off

If I understand correctly, there are two issues:

1. A MIDI device is not connected, GO starts, the MIDI device is connected, GO doesn't connect to the new MIDI port. A workaround is the Panic Button.
2. GO is running and is connected to a MIDI port, USB is disconnected and reconnected, GO doesn't connect to the new MIDI port.

Fixing issue 2 by reconnecting USB to the old MIDI port doesn't fix issue 1.
Fixing issue 1 does also fix issue 2, or did I miss something?
If RtMidi can do a simple "Setup Changed" callback or GO can detect changes in the setup then GO can do the MIDI part of the Panic button and both issues are fixed. How difficult would it be to implement this?

**Comment by Psychlist1972** [comment]
Yes. 1 is completely different from 2. 

1 is easily solved. For WinMM, it requires checking the port count from time to time (not in a tight loop like so many apps do. Every half second or so is much more than enough). For WinRT MIDI 1.0, a DeviceWatcher is the way to go, but IIRC, rtMIDI only takes a snapshot on startup instead of responding to device add/remove events. For Windows MIDI Services, there's a new version of the Device Watcher, but it works the same way.

Issue 1 is an issue around port enumeration.

Issue 2 is an issue around in-use-connection reconstruction/reconnection.

Pete
Microsoft

**Comment by midimusic** [comment]
@Psychlist1972

Thank you for your very constructive contribution.
Are you now able to point to some example code (C or C++) which consolidates all this into a practical solution for programming midi ports in Windows which will automatically and reliably re-connect? Particularly with regard to coping with ports which have the same name and future compatability with MIDI V2.0. To be most useful it must also work with older versions of Windows, ideally  back to XP.

csw900

**Comment by madbrain76** [comment]
Sorry for the late update here. I found actually that my rtmidi PR only fixes half of my issue in GO.

If I start GO with my MIDI interface powered on, the rtmidi fix allows subsequently powering off the interface, and back on, and this is completely transparent for GO, and presumably for other rtmidi apps as well.

However, if GO starts with the MIDI interface disconnected, GO never opens it at startup. And thus the rtmidi auto-reconnect patch does not fix this case. Unfortunately, that requires a change at the application level, ie. in GO. I have asked Codex to come up with a fix for this as well. I just tested it, and it works, but it is not ready for a PR submission yet.

My rtmidi & GO changes are separate, and don't depend on each other.

**Comment by madbrain76** [comment]
My changes for rtmidi were merged. I submitted a PR to pick up the latest version in Grandorgue : https://github.com/GrandOrgue/grandorgue/pull/2644

**Comment by madbrain76** [comment]
I submitted a PR for late MIDI device discovery. This allows Grandorgue to open a MIDI device that was previously configured, but absent as program startup. This is a cross-platform, and should be straightforward to review. I have only tested it on Linux with Alsa, however, so it could benefit from other platform testing.

https://github.com/GrandOrgue/grandorgue/pull/2645/changes/8420132e82328b86e427bfccf012df0c6cc985de

Note that subsequent auto-reconnect behavior is dependent on the combination of rtmidi backend  and OS.

This is the current status for auto-reconnect by backend/OS, according to Codex.

<img width="3794" height="996" alt="image" src="https://github.com/user-attachments/assets/1a20124d-2f5c-4fff-8ce2-e573b12174de" />

**Reply by willeke1234** [reply]
I’ve already done some testing on macOS, but I’m waiting for a response from Oleg first. Is there any documentation on how it is supposed to work?
Btw, is the reconnection overview only for USB?

**Reply by madbrain76** [reply]
There are multiple fixes, which one are you testing ? Or are you testing all of them ?

For Mac, this is the PR you want to test :
https://github.com/GrandOrgue/grandorgue/pull/2645

This one not USB-specific. The test is simple : disconnect your MIDI device, start Grandorgue, then plug your MIDI device. Start using your MIDI device. It should automatically work, without needing to go back into settings and validate again.

For auto-reconnection, on Mac, the auto-reconnection logic should already have worked, and behavior should be unchanged. 
This https://github.com/GrandOrgue/grandorgue/pull/2644 is for USB, and provides the same behavior on Alsa/Linux, but will have to effect on Mac or any other platform.

I also have a pair of old Echo Firewire 8a & 12 audio DAC/ADC/mixers with MIDI interfaces I'm going to try next. I added a Firewire PCIe card to my Proxmox box and will be testing that case under Linux as well. A few of my projects are to create proper mixers for these, as well as a GUI patchbay interface for my UM-880.

**Reply by willeke1234** [reply]
I'm testing #2645 without #2644.

macOS auto-reconnects devices (USB, BLE), but it doesn't reconnect virtual connections.
